File: apps/qt-ai-cli/src/commands/retry.ts
import { existsSync, renameSync, rmSync } from "node:fs";
import { join, resolve } from "node:path";
import { loadState, saveState, storyPaths, workFile, type StoryPaths, type StoryState, type WorkKind } from "../story-fs.ts";
import { naturalChapterCompare } from "@/lib/ai-story";

const WORK_KINDS: WorkKind[] = ["prompt", "draft", "glossary", "check", "review"];

/** Bản sao lưu của out/<id>.txt khi dịch lại chương đã done (đè bak cũ, không tích luỹ). */
export function retryBackupPath(paths: StoryPaths, id: string): string {
  return join(paths.outDir, `${id}.txt.bak`);
}

/**
 * Đưa chương về hàng đợi dịch lại từ đầu (reviewRound 0, dọn work/).
 * - error/skipped: ca "cứu chương hỏng" thông thường.
 * - done: out/<id>.txt đổi tên thành .txt.bak để lỡ tay còn lấy lại; accept lần sau ghi bản mới.
 * - translating: chương kẹt sau khi phiên chết — caller tự chắc không có phiên đang chạy trên truyện.
 * - queued: từ chối, chương đã ở hàng đợi.
 */
export function runRetry(root: string, id: string): void {
  const paths = storyPaths(resolve(root));
  const state = loadState(paths);
  const chapter = state.chapters[id];
  if (!chapter) throw new Error(`Không có chương ${id} trong state.json.`);
  if (chapter.status === "queued") throw new Error(`Chương ${id} đang queued sẵn rồi.`);
  requeueChapter(paths, state, id);
  saveState(paths, state);
}

export interface RetryRangeOutcome {
  retried: string[];
  backedUp: string[];
  alreadyQueued: string[];
}

/** Dịch lại mọi chương trong [from..to] (undefined = đầu/cuối): done giữ .bak, queued bỏ qua. */
export function runRetryRange(root: string, from?: string, to?: string): RetryRangeOutcome {
  const paths = storyPaths(resolve(root));
  const state = loadState(paths);
  const all = Object.keys(state.chapters).sort(naturalChapterCompare);
  for (const bound of [from, to]) {
    if (bound !== undefined && !state.chapters[bound]) throw new Error(`Không có chương ${bound} trong state.json.`);
  }
  if (from !== undefined && to !== undefined && naturalChapterCompare(from, to) > 0) {
    throw new Error(`Khoảng ngược: --from ${from} sau --to ${to}.`);
  }
  const outcome: RetryRangeOutcome = { retried: [], backedUp: [], alreadyQueued: [] };
  for (const id of all) {
    if (from !== undefined && naturalChapterCompare(id, from) < 0) continue;
    if (to !== undefined && naturalChapterCompare(id, to) > 0) continue;
    const status = state.chapters[id]!.status;
    if (status === "queued") {
      outcome.alreadyQueued.push(id);
      continue;
    }
    if (status === "done") outcome.backedUp.push(id);
    requeueChapter(paths, state, id);
    outcome.retried.push(id);
  }
  if (outcome.retried.length > 0) saveState(paths, state);
  return outcome;
}

/** Một chương về hàng đợi trong state (chưa ghi đĩa): done → out/<id>.txt thành .bak; dọn work/. */
function requeueChapter(paths: StoryPaths, state: StoryState, id: string): void {
  if (state.chapters[id]?.status === "done") {
    const out = join(paths.outDir, `${id}.txt`);
    if (existsSync(out)) renameSync(out, retryBackupPath(paths, id));
  }
  state.chapters[id] = { status: "queued", reviewRound: 0, updatedAt: Date.now() };
  for (const kind of WORK_KINDS) rmSync(workFile(paths, id, kind), { force: true });
}

File: crates/qt-ai-core/src/commands/retry.rs
use crate::error::{CoreError, Result};
use crate::story::natural_chapter_compare;
use crate::story_fs::{
    load_state, resolve_root, save_state, story_paths, work_file, ChapterState, ChapterStatus, StoryPaths, StoryState,
    WORK_KINDS,
};
use std::cmp::Ordering;
use std::fs;
use std::path::{Path, PathBuf};

/// Bản sao lưu của `out/<id>.txt` khi dịch lại chương đã done (đè bak cũ, không tích luỹ).
pub fn retry_backup_path(paths: &StoryPaths, id: &str) -> PathBuf {
    paths.out_dir.join(format!("{id}.txt.bak"))
}

/// Đưa chương về hàng đợi dịch lại từ đầu (reviewRound 0, dọn work/).
/// - error/skipped: ca "cứu chương hỏng" thông thường.
/// - done: `out/<id>.txt` đổi tên thành `.txt.bak` để lỡ tay còn lấy lại; accept lần sau ghi bản mới.
/// - translating: chương kẹt sau khi phiên chết — caller phải tự chắc không có phiên nào đang chạy
///   trên truyện này (GUI chặn khi phiên đang chạy), vì runner có thể đang ghi work/ của chính chương đó.
/// - queued: từ chối, chương đã ở hàng đợi.
pub fn run_retry(root: &Path, id: &str) -> Result<()> {
    let paths = story_paths(&resolve_root(root));
    let mut state = load_state(&paths)?;
    let chapter = state
        .chapters
        .get(id)
        .ok_or_else(|| CoreError::StoryNotFound(format!("Không có chương {id} trong state.json.")))?;
    if chapter.status == ChapterStatus::Queued {
        return Err(CoreError::InvalidState(format!("Chương {id} đang queued sẵn rồi.")));
    }
    requeue_chapter(&paths, &mut state, id)?;
    save_state(&paths, &state)
}

/// Kết quả dịch lại nhiều chương.
#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct RetryRangeOutcome {
    /// Chương đã đưa về hàng đợi, theo thứ tự tự nhiên.
    pub retried: Vec<String>,
    /// Chương done bị đổi out/<id>.txt thành .bak (tập con của `retried`).
    pub backed_up: Vec<String>,
    /// Chương vốn đã queued, bỏ qua.
    pub already_queued: Vec<String>,
}

/// Dịch lại mọi chương trong [from..to] (None = từ đầu / tới cuối), cùng luật với `run_retry` từng
/// chương: done → giữ .bak, queued bỏ qua. Caller (GUI/CLI) tự chắc không có phiên đang chạy.
pub fn run_retry_range(root: &Path, from: Option<&str>, to: Option<&str>) -> Result<RetryRangeOutcome> {
    let paths = story_paths(&resolve_root(root));
    let mut state = load_state(&paths)?;
    let mut all: Vec<String> = state.chapters.keys().cloned().collect();
    all.sort_by(|a, b| natural_chapter_compare(a, b));
    for bound in [from, to].into_iter().flatten() {
        if !state.chapters.contains_key(bound) {
            return Err(CoreError::StoryNotFound(format!("Không có chương {bound} trong state.json.")));
        }
    }
    if let (Some(from), Some(to)) = (from, to) {
        if natural_chapter_compare(from, to) == Ordering::Greater {
            return Err(CoreError::InvalidState(format!("Khoảng ngược: --from {from} sau --to {to}.")));
        }
    }
    let in_range = all.into_iter().filter(|id| {
        from.is_none_or(|from| natural_chapter_compare(id, from) != Ordering::Less)
            && to.is_none_or(|to| natural_chapter_compare(id, to) != Ordering::Greater)
    });
    let mut outcome = RetryRangeOutcome::default();
    for id in in_range {
        let status = state.chapters[&id].status;
        if status == ChapterStatus::Queued {
            outcome.already_queued.push(id);
            continue;
        }
        if status == ChapterStatus::Done {
            outcome.backed_up.push(id.clone());
        }
        requeue_chapter(&paths, &mut state, &id)?;
        outcome.retried.push(id);
    }
    if !outcome.retried.is_empty() {
        save_state(&paths, &state)?;
    }
    Ok(outcome)
}

/// Một chương về hàng đợi trong `state` (chưa ghi đĩa): done → out/<id>.txt thành .bak; dọn work/.
fn requeue_chapter(paths: &StoryPaths, state: &mut StoryState, id: &str) -> Result<()> {
    if state.chapters[id].status == ChapterStatus::Done {
        let out = paths.out_dir.join(format!("{id}.txt"));
        if out.is_file() {
            let backup = retry_backup_path(paths, id);
            fs::rename(&out, &backup).map_err(|source| CoreError::Io { path: out.clone(), source })?;
        }
    }
    state.chapters.insert(id.to_string(), ChapterState::fresh(ChapterStatus::Queued));
    for kind in WORK_KINDS {
        let _ = fs::remove_file(work_file(paths, id, kind));
    }
    Ok(())
}
